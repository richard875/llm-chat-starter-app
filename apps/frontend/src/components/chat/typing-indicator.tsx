export const TypingIndicator = () => {
  return (
    <div className="flex items-center space-x-1 px-2 py-1">
      <div className="h-2 w-2 animate-bounce rounded-full bg-primary [animation-delay:-0.3s]" />
      <div className="h-2 w-2 animate-bounce rounded-full bg-primary [animation-delay:-0.15s]" />
      <div className="h-2 w-2 animate-bounce rounded-full bg-primary" />
    </div>
  );
};
