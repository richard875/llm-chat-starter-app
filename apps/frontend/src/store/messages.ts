import { create } from "zustand";

import type { Message, Chat } from "@/types/chat";

interface MessagesState {
  messages: Message[];
  chats: Chat[];
  currentChatId: string | null;
  isLoadingChats: boolean;
  addMessage: (message: Message) => void;
  updateLastMessage: (content: string) => void;
  clearMessages: () => void;
  setCurrentChatId: (chatId: string) => void;
  startNewChat: () => void;
  fetchChats: () => Promise<void>;
  loadChatMessages: (chatId: string) => Promise<void>;
  refreshChatsAfterNewChat: () => Promise<void>;
}

export const useMessages = create<MessagesState>((set) => ({
  messages: [],
  chats: [],
  currentChatId: null,
  isLoadingChats: false,
  addMessage: (message) =>
    set((state) => ({ messages: [...state.messages, message] })),
  updateLastMessage: (content) =>
    set((state) => ({
      messages: state.messages.map((msg, idx) =>
        idx === state.messages.length - 1
          ? { ...msg, content: msg.content + content }
          : msg
      ),
    })),
  clearMessages: () => set({ messages: [] }),
  setCurrentChatId: (chatId) => set({ currentChatId: chatId }),
  startNewChat: () => {
    set({ messages: [], currentChatId: null });
  },
  fetchChats: async () => {
    set({ isLoadingChats: true });
    try {
      const response = await fetch("/api/chats");
      if (response.ok) {
        const data = await response.json();
        set({ chats: data.chats, isLoadingChats: false });
      } else {
        console.error("Failed to fetch chats");
        set({ isLoadingChats: false });
      }
    } catch (error) {
      console.error("Error fetching chats:", error);
      set({ isLoadingChats: false });
    }
  },
  loadChatMessages: async (chatId: string) => {
    try {
      const response = await fetch(`/api/messages/${chatId}`);
      if (response.ok) {
        const data = await response.json();
        set({ messages: data.messages, currentChatId: chatId });
      } else {
        console.error("Failed to fetch chat messages");
      }
    } catch (error) {
      console.error("Error fetching chat messages:", error);
    }
  },
  refreshChatsAfterNewChat: async () => {
    try {
      const response = await fetch("/api/chats");
      if (response.ok) {
        const data = await response.json();
        set({ chats: data.chats });
      }
    } catch (error) {
      console.error("Error refreshing chats:", error);
    }
  },
}));
