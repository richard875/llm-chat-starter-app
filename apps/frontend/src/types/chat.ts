export interface Message {
  id?: number;
  chatId?: string;
  role: "user" | "assistant";
  content: string;
  createdAt?: string;
  updatedAt?: string;
}

export interface Chat {
  chatId: string;
  title: string;
  createdAt?: string;
  updatedAt?: string;
}
