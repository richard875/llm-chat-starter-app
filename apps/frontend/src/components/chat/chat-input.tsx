import { SendHorizonal } from "lucide-react";
import { useEffect, useRef, useState } from "react";

import { Button } from "@/components/ui/button";
import { Input } from "@/components/ui/input";
import { useMessages } from "@/store/messages";
import { generateChatId } from "@/lib/utils";

interface ChatInputProps {
  onTypingChange: (isTyping: boolean) => void;
}

export const ChatInput = ({ onTypingChange }: ChatInputProps) => {
  const [input, setInput] = useState("");
  const {
    messages,
    isLoadingMsg,
    addMessage,
    updateLastMessage,
    currentChatId,
    setCurrentChatId,
    setIsLoadingMsg,
    refreshChatsAfterNewChat,
  } = useMessages();
  const inputRef = useRef<HTMLInputElement>(null);

  // Focus input when loading state changes to false
  useEffect(() => {
    if (!isLoadingMsg) {
      inputRef.current?.focus();
    }
    onTypingChange(isLoadingMsg);
  }, [isLoadingMsg, onTypingChange]);

  const handleSubmit = async (e: React.FormEvent) => {
    e.preventDefault();

    if (!input.trim() || isLoadingMsg) return;

    // Generate a new chat ID if this is the first message in a new chat
    const isNewChat = !currentChatId;
    const chatId = isNewChat ? generateChatId() : currentChatId;
    setCurrentChatId(chatId);

    // Add user message
    addMessage({
      role: "user",
      content: input,
      chatId,
    });

    // Add empty assistant message that will be streamed
    addMessage({
      role: "assistant",
      content: "",
      chatId,
    });

    setInput("");
    setIsLoadingMsg(true);

    try {
      const response = await fetch("/api/chat", {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          chatId,
          messages: [...messages, { role: "user", content: input }],
        }),
      });

      if (!response.ok) {
        throw new Error("Failed to get response");
      }

      const reader = response.body?.getReader();
      if (!reader) throw new Error("No reader available");

      const decoder = new TextDecoder();

      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        const chunk = decoder.decode(value);
        const lines = chunk.split("\n");

        for (const line of lines) {
          if (line.startsWith("data: ")) {
            try {
              const data = JSON.parse(line.slice(6));
              if (data.content) {
                updateLastMessage(data.content);
              }
            } catch (e) {
              console.error("Error parsing SSE data:", e);
            }
          }
        }
      }
    } catch (error) {
      console.error("Error:", error);
      updateLastMessage("Sorry, there was an error processing your request.");
    } finally {
      setIsLoadingMsg(false);
      // Refresh chats list if this was a new chat
      if (isNewChat) refreshChatsAfterNewChat();
    }
  };

  return (
    <form
      onSubmit={handleSubmit}
      className="flex w-full items-center space-x-3 rounded-full p-2 border border-primary shadow"
    >
      <Input
        ref={inputRef}
        value={input}
        onChange={(e) => setInput(e.target.value)}
        placeholder="Type your message..."
        name="message"
        className="flex-1 border-0 shadow-none !text-base focus-visible:ring-0"
        disabled={isLoadingMsg}
      />
      <Button
        type="submit"
        className="size-10 rounded-full bg-[#a3e636] border border-primary hover:bg-[#91cc33] cursor-pointer transition"
        disabled={isLoadingMsg}
      >
        <SendHorizonal className="text-primary" />
      </Button>
    </form>
  );
};
