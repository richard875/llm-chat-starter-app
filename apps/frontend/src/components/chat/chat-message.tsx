import ReactMarkdown from "react-markdown";

import { TypingIndicator } from "@/components/chat/typing-indicator";
import { cn } from "@/lib/utils";
import type { Message } from "@/types/chat";

interface ChatMessageProps {
  message: Message;
  isTyping?: boolean;
}

export const ChatMessage = ({ message, isTyping }: ChatMessageProps) => {
  const isUser = message.role === "user";

  return (
    <div className={cn("flex mb-5", isUser ? "justify-end" : "justify-start")}>
      <div
        className={cn(
          "px-5 py-3 rounded-3xl",
          isUser
            ? "bg-primary text-primary-foreground"
            : "text-foreground border-primary border-1"
        )}
      >
        {message.content ? (
          <div className="leading-6 text-base tracking-wide">
            <ReactMarkdown>{message.content}</ReactMarkdown>
          </div>
        ) : isTyping ? (
          <TypingIndicator />
        ) : null}
      </div>
    </div>
  );
};
