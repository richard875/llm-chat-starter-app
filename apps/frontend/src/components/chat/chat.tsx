import { useEffect, useRef, useState } from "react";

import { ChatInput } from "@/components/chat/chat-input";
import { ChatMessage } from "@/components/chat/chat-message";
import { useMessages } from "@/store/messages";

export const Chat = () => {
  const { messages } = useMessages();
  const scrollAreaRef = useRef<HTMLDivElement>(null);
  const [isTyping, setIsTyping] = useState(false);

  // Scroll to bottom when messages change or typing state changes
  useEffect(() => {
    if (scrollAreaRef.current) {
      scrollAreaRef.current.scrollTop = scrollAreaRef.current.scrollHeight;
    }
  }, [messages, isTyping]);

  return (
    <div className="flex flex-col h-full">
      <div className="flex-1 overflow-y-auto" ref={scrollAreaRef}>
        {messages.length === 0 ? (
          <div className="w-full h-full pb-10 flex items-center justify-center select-none">
            <p className="text-2xl font-medium">
              Start a conversation with the AI assistant.
            </p>
          </div>
        ) : (
          <div className="max-w-screen-md mx-auto py-8 px-4">
            <div className="space-y-4">
              {messages.map((message, index) => (
                <ChatMessage
                  key={index}
                  message={message}
                  isTyping={
                    isTyping &&
                    index === messages.length - 1 &&
                    message.role === "assistant"
                  }
                />
              ))}
            </div>
          </div>
        )}
      </div>
      <div className="w-full bg-background z-20 px-5">
        <div className="max-w-screen-md mx-auto">
          <ChatInput onTypingChange={setIsTyping} />
          <p className="text-sm text-center py-2.5 text-muted-foreground select-none">
            AI can make mistakes. Please verify the information provided.
          </p>
        </div>
      </div>
    </div>
  );
};
