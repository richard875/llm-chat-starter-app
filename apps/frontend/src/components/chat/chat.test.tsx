import { describe, test, expect, vi, beforeEach } from "vitest";
import { render } from "@testing-library/react";
import { Chat } from "@/components/chat/chat";
import { useMessages } from "@/store/messages";
import type { Message } from "@/types/chat";

vi.mock("@/store/messages");

describe("Chat", () => {
  const mockUseMessages = vi.mocked(useMessages);

  beforeEach(() => {
    vi.clearAllMocks();
    // Default mock implementation
    mockUseMessages.mockReturnValue({
      messages: [],
      currentChatId: null,
      isLoadingMsg: false,
      addMessage: vi.fn(),
      updateLastMessage: vi.fn(),
      setCurrentChatId: vi.fn(),
      setIsLoadingMsg: vi.fn(),
      refreshChatsAfterNewChat: vi.fn(),
      isLoadingChats: false,
      chats: [],
      startNewChat: vi.fn(),
      fetchChats: vi.fn(),
      loadChatMessages: vi.fn(),
    });
  });

  test("renders empty state when no messages", () => {
    const { container } = render(<Chat />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test("renders with messages", () => {
    const mockMessages: Message[] = [
      {
        id: 1,
        role: "user",
        content: "Hello",
        chatId: "chat-1",
        createdAt: "2023-01-01T00:00:00.000Z",
      },
      {
        id: 2,
        role: "assistant",
        content: "Hi there!",
        chatId: "chat-1",
        createdAt: "2023-01-01T00:01:00.000Z",
      },
    ];

    mockUseMessages.mockReturnValue({
      messages: mockMessages,
      currentChatId: "chat-1",
      isLoadingMsg: false,
      addMessage: vi.fn(),
      updateLastMessage: vi.fn(),
      setCurrentChatId: vi.fn(),
      setIsLoadingMsg: vi.fn(),
      refreshChatsAfterNewChat: vi.fn(),
      isLoadingChats: false,
      chats: [],
      startNewChat: vi.fn(),
      fetchChats: vi.fn(),
      loadChatMessages: vi.fn(),
    });

    const { container } = render(<Chat />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test("renders with current chat ID set", () => {
    mockUseMessages.mockReturnValue({
      messages: [],
      currentChatId: "active-chat-123",
      isLoadingMsg: false,
      addMessage: vi.fn(),
      updateLastMessage: vi.fn(),
      setCurrentChatId: vi.fn(),
      setIsLoadingMsg: vi.fn(),
      refreshChatsAfterNewChat: vi.fn(),
      isLoadingChats: false,
      chats: [],
      startNewChat: vi.fn(),
      fetchChats: vi.fn(),
      loadChatMessages: vi.fn(),
    });

    const { container } = render(<Chat />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
