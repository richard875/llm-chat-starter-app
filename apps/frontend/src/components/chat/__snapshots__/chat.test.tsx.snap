// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Chat > renders empty state when no messages 1`] = `
<div
  class="flex-1 flex flex-col h-full"
>
  <div
    class="flex-1 overflow-y-auto"
  >
    <div
      animate="[object Object]"
      class="w-full h-full pb-10 flex items-center justify-center select-none"
      exit="[object Object]"
      initial="[object Object]"
      transition="[object Object]"
    >
      <p
        class="text-2xl font-medium"
      >
        Start a conversation with the AI assistant.
      </p>
    </div>
  </div>
  <div
    class="w-full bg-background z-20 px-5"
  >
    <div
      class="max-w-screen-md mx-auto"
    >
      <form
        class="flex w-full items-center rounded-full p-2 border border-primary shadow"
      >
        <input
          class="file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-input flex h-9 w-full min-w-0 rounded-md bg-transparent px-3 py-1 text-base transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive flex-1 border-0 shadow-none !text-base focus-visible:ring-0 select-none"
          data-slot="input"
          name="message"
          placeholder="Type your message..."
          value=""
        />
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive text-primary-foreground shadow-xs px-4 py-2 has-[>svg]:px-3 size-10 rounded-full border border-primary cursor-pointer transition bg-[#a3e636] hover:bg-[#91cc33]"
          data-slot="button"
          type="submit"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-send-horizontal text-primary"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3.714 3.048a.498.498 0 0 0-.683.627l2.843 7.627a2 2 0 0 1 0 1.396l-2.842 7.627a.498.498 0 0 0 .682.627l18-8.5a.5.5 0 0 0 0-.904z"
            />
            <path
              d="M6 12h16"
            />
          </svg>
        </button>
      </form>
      <p
        class="text-sm text-center py-2.5 text-muted-foreground select-none"
      >
        <span
          class="hidden md:inline-block"
        >
          AI can make mistakes. 
        </span>
        Please verify the information provided.
      </p>
    </div>
  </div>
</div>
`;

exports[`Chat > renders with current chat ID set 1`] = `
<div
  class="flex-1 flex flex-col h-full"
>
  <div
    class="flex-1 overflow-y-auto"
  >
    <div
      animate="[object Object]"
      class="w-full h-full pb-10 flex items-center justify-center select-none"
      exit="[object Object]"
      initial="[object Object]"
      transition="[object Object]"
    >
      <p
        class="text-2xl font-medium"
      >
        Start a conversation with the AI assistant.
      </p>
    </div>
  </div>
  <div
    class="w-full bg-background z-20 px-5"
  >
    <div
      class="max-w-screen-md mx-auto"
    >
      <form
        class="flex w-full items-center rounded-full p-2 border border-primary shadow"
      >
        <input
          class="file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-input flex h-9 w-full min-w-0 rounded-md bg-transparent px-3 py-1 text-base transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive flex-1 border-0 shadow-none !text-base focus-visible:ring-0 select-none"
          data-slot="input"
          name="message"
          placeholder="Type your message..."
          value=""
        />
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive text-primary-foreground shadow-xs px-4 py-2 has-[>svg]:px-3 size-10 rounded-full border border-primary cursor-pointer transition bg-[#a3e636] hover:bg-[#91cc33]"
          data-slot="button"
          type="submit"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-send-horizontal text-primary"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3.714 3.048a.498.498 0 0 0-.683.627l2.843 7.627a2 2 0 0 1 0 1.396l-2.842 7.627a.498.498 0 0 0 .682.627l18-8.5a.5.5 0 0 0 0-.904z"
            />
            <path
              d="M6 12h16"
            />
          </svg>
        </button>
      </form>
      <p
        class="text-sm text-center py-2.5 text-muted-foreground select-none"
      >
        <span
          class="hidden md:inline-block"
        >
          AI can make mistakes. 
        </span>
        Please verify the information provided.
      </p>
    </div>
  </div>
</div>
`;

exports[`Chat > renders with messages 1`] = `
<div
  class="flex-1 flex flex-col h-full"
>
  <div
    class="flex-1 overflow-y-auto"
  >
    <div
      animate="[object Object]"
      class="w-full bg-background z-20 px-5 py-8"
      exit="[object Object]"
      initial="[object Object]"
      transition="[object Object]"
    >
      <div
        class="max-w-screen-md mx-auto"
      >
        <div
          animate="[object Object]"
          exit="[object Object]"
          initial="[object Object]"
          transition="[object Object]"
        >
          <div
            class="flex mb-7 last:mb-0 justify-end"
          >
            <div
              class="min-w-0 max-w-full bg-primary text-primary-foreground px-4 py-2 rounded-[20px]"
            >
              <div
                class="leading-6.25 text-[15px] tracking-wide min-w-0 w-full"
              >
                <p
                  class="mb-4 last:mb-0"
                >
                  Hello
                </p>
              </div>
            </div>
          </div>
          <div
            class="flex mb-7 last:mb-0 justify-start"
          >
            <div
              class="min-w-0 max-w-full text-foreground"
            >
              <div
                class="leading-6.25 text-[15px] tracking-wide min-w-0 w-full"
              >
                <p
                  class="mb-4 last:mb-0"
                >
                  Hi there!
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="w-full bg-background z-20 px-5"
  >
    <div
      class="max-w-screen-md mx-auto"
    >
      <form
        class="flex w-full items-center rounded-full p-2 border border-primary shadow"
      >
        <input
          class="file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-input flex h-9 w-full min-w-0 rounded-md bg-transparent px-3 py-1 text-base transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive flex-1 border-0 shadow-none !text-base focus-visible:ring-0 select-none"
          data-slot="input"
          name="message"
          placeholder="Type your message..."
          value=""
        />
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive text-primary-foreground shadow-xs px-4 py-2 has-[>svg]:px-3 size-10 rounded-full border border-primary cursor-pointer transition bg-[#a3e636] hover:bg-[#91cc33]"
          data-slot="button"
          type="submit"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-send-horizontal text-primary"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3.714 3.048a.498.498 0 0 0-.683.627l2.843 7.627a2 2 0 0 1 0 1.396l-2.842 7.627a.498.498 0 0 0 .682.627l18-8.5a.5.5 0 0 0 0-.904z"
            />
            <path
              d="M6 12h16"
            />
          </svg>
        </button>
      </form>
      <p
        class="text-sm text-center py-2.5 text-muted-foreground select-none"
      >
        <span
          class="hidden md:inline-block"
        >
          AI can make mistakes. 
        </span>
        Please verify the information provided.
      </p>
    </div>
  </div>
</div>
`;
